dbms/src/Flash/Mpp/GRPCReceiveQueue.h

```diff
@@ -0,0 +1,250 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/ConcurrentIOQueue.h>
+#include <Common/Exception.h>
+#include <Common/Logger.h>
+#include <Common/grpcpp.h>
+#include <common/logger_useful.h>
+
+#include <atomic>
+#include <functional>
+#include <magic_enum.hpp>
+#include <mutex>
+#include <queue>
+#include <set>
+#include <utility>
+
+namespace DB
+{
+namespace tests
+{
+class TestGRPCReceiveQueue;
+} // namespace tests
+
+/// In grpc cpp framework, the tag that is pushed into grpc completion
+/// queue must be inherited from `CompletionQueueTag`.
+class KickReceiveTag : public grpc::internal::CompletionQueueTag
+{
+public:
+    explicit KickReceiveTag(void * tag_)
+        : tag(tag_)
+    {}
+
+    bool FinalizeResult(void ** tag_, bool * /*status*/) override
+    {
+        *tag_ = tag;
+        return true;
+    }
+
+private:
+    void * tag;
+};
+
+using GRPCReceiveKickFunc = std::function<grpc_call_error(KickReceiveTag *, grpc_call *)>;
+using AsyncRetryConnection = std::pair<KickReceiveTag *, grpc_call *>;
+
+// This is a FIFO data struct.
+//
+// When MPMCQueue is full, put the information related with AsyncRequestHandler
+// into this queue so that we can kick it into GRPCCompletionQueue when the
+// MPMCQueue is not full and try to re-write the data.
+class AsyncRequestHandlerWaitQueue
+{
+public:
+    bool empty()
+    {
+        std::lock_guard lock(mu);
+        return wait_retry_queue.empty();
+    }
+
+    void push(AsyncRetryConnection conn)
+    {
+        std::lock_guard lock(mu);
+        wait_retry_queue.push(conn);
+    }
+
+    AsyncRetryConnection pop()
+    {
+        std::lock_guard lock(mu);
+        if (wait_retry_queue.empty())
+            return AsyncRetryConnection(nullptr, nullptr);
+
+        AsyncRetryConnection ret_conn = wait_retry_queue.front();
+        wait_retry_queue.pop();
+
+        return ret_conn;
+    }
+
+    std::queue<AsyncRetryConnection> popAll()
+    {
+        std::lock_guard lock(mu);
+        if (wait_retry_queue.empty())
+            return std::queue<AsyncRetryConnection>{};
+
+        std::queue<AsyncRetryConnection> ret_conn = std::move(wait_retry_queue);
+        return ret_conn;
```

Comment:
I am a little worried about whether using move will cause problems with `wait_retry_queue`.
```suggestion
        std::queue<AsyncRetryConnection> ret;
        {
            std::lock_guard lock(mu);
            std::swap(ret, wait_retry_queue);
        }
        return wait_retry_queue;
```

Reply:
> I am a little worried about whether using move will cause problems with `wait_retry_queue`.



> I am a little worried about whether using move will cause problems with `wait_retry_queue`.

Repeatedly using a moved object seems to violate the semantic, I will change it.